waspc/src/Wasp/Generator/DbGenerator.hs

```diff
@@ -99,7 +110,49 @@ genPrismaSchema wasp = createTemplateFileDraft dstPath tmplSrcPath (Just templat
 
 genMigrationsDir :: Wasp -> Maybe FileDraft
 genMigrationsDir wasp =
-  (getMigrationsDir wasp) >>= \waspMigrationsDir ->
+  getMigrationsDir wasp >>= \waspMigrationsDir ->
     Just $ createCopyDirFileDraft (SP.castDir genProjectMigrationsDir) (SP.castDir waspMigrationsDir)
   where
     genProjectMigrationsDir = dbRootDirInProjectRootDir </> dbMigrationsDirInDbRootDir
+
+writeDbSchemaChecksumToFile :: Path' Abs (Dir ProjectRootDir) -> IO ()
+writeDbSchemaChecksumToFile genProjectRootDir = do
+  dbSchemaExists <- doesFileExist dbSchemaFp
+  when dbSchemaExists $ do
+    checksum <- calculateChecksumStringForFile dbSchemaFp
+    writeFile dbSchemaChecksumFp checksum
+  where
+    dbSchemaFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaFileInProjectRootDir
+    dbSchemaChecksumFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaChecksumFileInProjectRootDir
+
+checkIfDbSchemaMatchesChecksum :: Path' Abs (Dir ProjectRootDir) -> ExceptT String IO ()
+checkIfDbSchemaMatchesChecksum genProjectRootDir = do
+  dbSchemaExists <- liftIO $ doesFileExist dbSchemaFp
+  let maybeDbSchemaFp = if dbSchemaExists then Just dbSchemaFp else Nothing
+  dbSchemaChecksumExists <- liftIO $ doesFileExist dbSchemaChecksumFp
+  let maybeDbSchemaChecksumFp = if dbSchemaChecksumExists then Just dbSchemaChecksumFp else Nothing
+  checkIfDbSchemaMatchesChecksum' maybeDbSchemaFp maybeDbSchemaChecksumFp
+  where
+    dbSchemaFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaFileInProjectRootDir
+    dbSchemaChecksumFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaChecksumFileInProjectRootDir
+
+checkIfDbSchemaMatchesChecksum' :: Maybe FilePath -> Maybe FilePath -> ExceptT String IO ()
+checkIfDbSchemaMatchesChecksum' (Just dbSchemaFp) (Just dbSchemaChecksumFp) = do
+  dbSchemaFileChecksum <- liftIO $ calculateChecksumStringForFile dbSchemaFp
+  dbChecksumFileContents <- liftIO $ readFile dbSchemaChecksumFp
+  if dbSchemaFileChecksum == dbChecksumFileContents
+    then return ()
+    else throwError "Your Prisma schema has changed, you should run wasp db migrate-dev."
+checkIfDbSchemaMatchesChecksum' (Just _dbSchemaFp) Nothing =
+  -- Cannot throw the error below because they may have no entities in the schema.prisma file.
+  -- throwError "You are missing the Prisma schema checksum. Please run wasp db migrate-dev."
+  return ()
+checkIfDbSchemaMatchesChecksum' Nothing (Just _dbSchemaChecksumFp) =
+  throwError "Your .wasp output directory appears malformed. Please clean by running wasp clean."
```

Comment:
I feel like if this situation happens, I am not even sure if they should run `wasp clean` or not hm. I mean it can't hurt, but maybe they don't have to do it? So maybe they deleted all entities from their project, and there is a leftover checksum file. Ok. But then they also have redundant migrations in their source project. They should also now to delete those. Hm. I guess it should be up to us to delete the checksum file if there is no schema.prisma, not to ask them to clean it up.

Reply:
There should always be a non-empty schema file (even if no entities). I'm honestly not sure how they would even get into this state, but figured "clean" is our universal reset so it was the safest suggestion here.